internal/forms/forms.go

```diff
@@ -0,0 +1,824 @@
+	retVal.Subject = strings.TrimSpace(retVal.Subject)
+	retVal.Body = strings.TrimSpace(retVal.Body)
+	return retVal, nil
+}
+
+func (b formMessageBuilder) initFormValues() {
+	if b.IsReply {
+		b.FormValues["msgisreply"] = "True"
+	} else {
+		b.FormValues["msgisreply"] = "False"
+	}
+
+	b.FormValues["msgsender"] = b.FormsMgr.config.MyCall
+
+	// some defaults that we can't set yet. Winlink doesn't seem to care about these
+	b.FormValues["msgto"] = ""
+	b.FormValues["msgcc"] = ""
+	b.FormValues["msgsubject"] = ""
+	b.FormValues["msgbody"] = ""
+	b.FormValues["msgp2p"] = ""
+	b.FormValues["msgisforward"] = "False"
+	b.FormValues["msgisacknowledgement"] = "False"
+	b.FormValues["msgseqnum"] = "0"
+}
+
+func (b formMessageBuilder) scanTmplBuildMessage(tmplPath string) (MessageForm, error) {
+
+	infile, err := os.Open(tmplPath)
+	if err != nil {
+		return MessageForm{}, err
+	}
+	defer infile.Close()
+
+	placeholderRegEx := regexp.MustCompile(`<[vV][aA][rR]\s+(\w+)\s*>`)
+	scanner := bufio.NewScanner(infile)
+
+	var retVal MessageForm
+	for scanner.Scan() {
+		lineTmpl := scanner.Text()
+		lineTmpl = fillPlaceholders(lineTmpl, placeholderRegEx, b.FormValues)
+		lineTmpl = strings.Replace(lineTmpl, "<MsgSender>", b.FormsMgr.config.MyCall, -1)
+		lineTmpl = strings.Replace(lineTmpl, "<ProgramVersion>", "Pat "+b.FormsMgr.config.AppVersion, -1)
+		if strings.HasPrefix(lineTmpl, "Form:") ||
+			strings.HasPrefix(lineTmpl, "ReplyTemplate:") ||
+			strings.HasPrefix(lineTmpl, "To:") ||
+			strings.HasPrefix(lineTmpl, "Msg:") {
+			continue
+		}
+		if b.Interactive {
+			matches := placeholderRegEx.FindAllStringSubmatch(lineTmpl, -1)
+			fmt.Println(string(lineTmpl))
+			for i := range matches {
+				varName := matches[i][1]
+				varNameLower := strings.ToLower(varName)
+				if b.FormValues[varNameLower] != "" {
+					continue
+				}
+				fmt.Print(varName + ": ")
+				b.FormValues[varNameLower] = "blank"
+				val := b.FormsMgr.config.LineReader()
+				if val != "" {
+					b.FormValues[varNameLower] = val
+				}
+			}
+		}
+
+		lineTmpl = fillPlaceholders(lineTmpl, placeholderRegEx, b.FormValues)
+		if strings.HasPrefix(lineTmpl, "Subject:") {
+			retVal.Subject = strings.TrimPrefix(lineTmpl, "Subject:")
+		} else {
+			retVal.Body += lineTmpl + "\n"
+		}
+	}
+
+	return retVal, nil
+}
+
+func xmlEscape(s string) string {
+	var buf bytes.Buffer
+	if err := xml.EscapeText(&buf, []byte(s)); err != nil {
+		log.Printf("Error trying to escape XML string %s", err)
+	}
+	return buf.String()
+}
+
+func fillPlaceholders(s string, re *regexp.Regexp, values map[string]string) string {
+	if _, ok := values["txtstr"]; !ok {
+		values["txtstr"] = ""
+	}
+	result := s
+	matches := re.FindAllStringSubmatch(s, -1)
+	for _, match := range matches {
+		value, ok := values[strings.ToLower(match[1])]
+		if ok {
+			result = strings.Replace(result, match[0], value, -1)
+		}
+	}
+	return result
+}
+
+func (mgr Manager) cleanupOldFormData() {
```

Comment:
@martinhpedersen, I am getting a warning from 'vet': "cleanupFormData passes lock by value". 
Same for all other member functions on Manager. I need help resolving those.

Reply:
Have a look at this blog post: https://medium.com/golangspec/detect-locks-passed-by-value-in-go-efb4ac9a3f2b

You're passing the Manager struct by value, which means that the Manager fields are copied for each method call. In cases where you want the Manager to be mutable (locking mutates it's state), it is common practice to pass by pointer intead:
```
func (m Manager) fooBar() {
  // m is a copy in this scope
  m.mu.Lock() // This locks a local copy of the mutex
}

func (m *Manager) fooBar() {
  // m is a pointer to a Manager in this scope
  m.mu.Lock() // This locks the shared mutex
}
```

I think the best solution is to change all the Manager methods to take a pointer receiver. `func (mgr Manager) foo()` -> `func (mgr *Manager) foo()`.

PS (off topic): It is also more idiomatic to use a single letter variable name for the receiver. Usually the first letter in of the type: `func (m *Manager) foo()`.

Reply:
More on the same topic here: https://eli.thegreenplace.net/2018/beware-of-copying-mutexes-in-go/

Reply:
fixed in 06cf2edb3d7f812f0eaff3aeaf60985b2e347506 and 3ac3f0c3248c21f5a98a6d5251e9a8afebe85877